curator-recipes/src/test/java/org/apache/curator/framework/recipes/nodes/TestPersistentTtlNode.java

```diff
@@ -237,4 +241,88 @@ void touch() {
             }
         }
     }
+
+    @Test
+    public void testInternalExecutorClose() throws Exception {
+        final String mainPath = "/parent/main";
+        final String touchPath = ZKPaths.makePath(mainPath, PersistentTtlNode.DEFAULT_CHILD_NODE_NAME);
+        final CountDownLatch touchCreatedLatch = new CountDownLatch(1);
+        try (CuratorFramework client =
+                CuratorFrameworkFactory.newClient(server.getConnectString(), new RetryOneTime(1))) {
+            client.start();
+            assertTrue(client.blockUntilConnected(1, TimeUnit.SECONDS));
+            try (PersistentWatcher watcher = new PersistentWatcher(client, mainPath, true)) {
+                final Watcher listener = event -> {
+                    final String path = event.getPath();
+                    if (touchPath.equals(path)) {
+                        touchCreatedLatch.countDown();
+                    }
+                };
+                watcher.getListenable().addListener(listener);
+                watcher.start();
+                final AtomicLong executorThreadId = new AtomicLong();
+                try (PersistentTtlNode node = new PersistentTtlNode(client, mainPath, ttlMs, new byte[0])) {
+                    node.start();
+                    assertTrue(touchCreatedLatch.await(5 * ttlMs, TimeUnit.MILLISECONDS));
+                    node.getCloseableScheduledExecutorService()
+                            .submit(() ->
+                                    executorThreadId.set(Thread.currentThread().getId()));
+                    assertFalse(getThreadsWithIdAndName(executorThreadId, PersistentTtlNode.TOUCH_THREAD_NAME)
+                            .isEmpty());
+                }
+                Thread.sleep(10L);
```

Comment:
No sure 10ms it is enough in testing env. How about export underlying executor for tests to `awaitTermination` ?

Reply:
I have my doubts here what is best, because awaitTermination assures you that tasks were executed, BUT the Javadoc does NOT confirm that the thread has been destroyed, so we can only rely that will happen eventually

After closing the PersistentTTLNode:
- When executor is internal, probably best is a loop with small sleeps to check that thread disappears. Notice also that CloseExecutorService is in another package and extracting methods for testing requires them to be public that I really do NOT like.
- When executor is provided, probably the best is to use awaitTermination and after that sleep for instance 100 msecs and check that the thread is still there

What do you think?

Reply:
Both sound good to me!

Reply:
I realized (doing the code) that awaitTermination requires that shutdown/shutdownNow has been previously invoked , otherwise will NOT work as expected . So when executor is provided, after PersistentTTLNode has been closed I will wait for 100ms and check that executor there is still there

Maybe looking for the thread using the threadId and the name is a bit too much, when it is enough to check if the executor isShutdown or NOT.

I will provide commit that is able to check if excutor is shutdown in both cases, and also the thread checks and you provide comments. Now I am more in favour to remove all the code that looks for the thread from stacks traces, BUT just give me your opinion there